### ISSUE TITLE
Implement actual minibatch learning

### ISSUE THREAD

**Issue by dzeber** [mozilla/CANOSP-2019#30]
The federated client update algorithm calls for splitting a client's data into "minibatches" of size B and updating the weights once per minibatch. The SGDClassifier doesn't in fact support this (there is an [open issue](https://github.com/scikit-learn/scikit-learn/issues/14468)) - it updates the weights once for each _training example_, which is what is traditionally meant by SGD, "stochastic" because it traverses the dataset in a random order (when `shuffle=True`).

Using `partial_fit` on a batch still causes one weight update per training example in the batch, so this isn't really getting us minibatch learning. The difference with `fit` is that it starts from the current state of the model and runs 1 weight update for each example in the supplied dataset, which may be different from the "main" dataset.

However, we can compute the minibatch update by averaging the pointwise updates across the minibatch. Given a minibatch of data x = (x_1,...,x_k), we want to compute w = w_0 + \eta \dell E(w_0, x). But since the [error](https://scikit-learn.org/stable/modules/sgd.html#mathematical-formulation) on the dataset E is the average of the error on each individual training example, E(w, x) = 1\n \sum_i E(w, x_i), we get the same result by computing pointwise updates w^{(i)} = w_0 + \eta \dell E(w_0, x_i) (as is done [at each step](https://scikit-learn.org/stable/modules/sgd.html#id1) by the SGDClassifier) and averaging the w^{(i)}: w = 1/n \sum_i w^{(i)}. The difference is that the built-in method updates the weights progressively, whereas we want the initial w_0 to be the same for each i.

I think this can be done by using a similar approach to the current version of `client_update`, _for each training example in the minibatch_ resetting the weights to the initial weights on entering the minibatch and running `partial_fit` on that training example. We would then average the weights at the end of the minibatch. We may want to create a fresh classifier instance each time rather than keep reusing the same one, because if we reuse, the internal step counter will keep incrementing. We should check the source code to see whether that will have any unintended consequences.

One outstanding issue is the choice of learning rate \eta. The default choice changes on each "time step", ie each per-example weight update. Let's pass this through as a param we handle. We can keep it constant for now, and maybe implement something adaptive later on.

**Comment by dzeber** [comment]
In progress in #43
